sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetScanBuilder.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2.parquet
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex
+import org.apache.spark.sql.execution.datasources.v2.FileScanBuilder
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.sources.v2.reader.{Scan, SupportsPushDownFilters}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+case class ParquetScanBuilder(
+    sparkSession: SparkSession,
+    fileIndex: PartitioningAwareFileIndex,
+    schema: StructType,
+    dataSchema: StructType,
+    options: CaseInsensitiveStringMap)
+  extends FileScanBuilder(sparkSession, fileIndex, dataSchema) with SupportsPushDownFilters {
+  lazy val hadoopConf = {
+    val caseSensitiveMap = options.asCaseSensitiveMap.asScala.toMap
+    // Hadoop Configurations are case sensitive.
+    sparkSession.sessionState.newHadoopConfWithOptions(caseSensitiveMap)
+  }
+
+  private var filters: Array[Filter] = _
+
+  override def pushFilters(filters: Array[Filter]): Array[Filter] = {
+    this.filters = filters
+    this.filters
+  }
+
+  // The actual filter push down happens in [[ParquetPartitionReaderFactory]].
+  // It requires the Parquet physical schema to determine whether a filter is convertible.
+  // So here we simply mark that all the filters are pushed down.
+  override def pushedFilters(): Array[Filter] = this.filters
```

Comment:
Sorry, but this looks like a violation of the contact of `SupportsPushDownFilters ` interface. Personally, I'm -1 for this kind of abuse because this Parquet code will be used as a reference model in the community for all the 3rd party data sources. Do we really need to allow this in Apache Spark code, @gengliangwang , @cloud-fan , @rxin , @rdblue , @dbtsai ?

Reply:
As per the discussion in https://github.com/apache/spark/pull/21696#pullrequestreview-133842990, we need to push down filters according to the actual physical schema in read tasks.

With the current Data Source V2 API, I think this is the simplest way we can make it happen. I understand this is ugly. But I can't think of other better solution here for now.

Also, in V1 (`FileSourceScanExec`), all the data filters without subqueries are considered as "pushed down" as well.

Reply:
I agree with @dongjoon-hyun, this should not violate the method contract.

Why is the file schema needed to push the filters? Iceberg pushes Parquet filters without knowing the physical schema of all the files. Can't this just return the filters that can be represented as Parquet filters?

Reply:
@dongjoon-hyun @rdblue I see. I will correct this. Thanks.

Reply:
Thank you, @gengliangwang !

Reply:
created https://github.com/apache/spark/pull/24597 for this.